Simplenote/src/main/java/com/automattic/simplenote/models/Tag.java

```diff
@@ -79,54 +79,59 @@ public void setIndex(Integer tagIndex) {
-            List<String> tagsHash = new ArrayList<>();
-
-            // Create lists of note's tags excluding old tag.
-            for (String tag : note.getTags()) {
-                if (!tag.equals(tagOld)) {
-                    tagsNew.add(tag);
-                    tagsHash.add(TagUtils.hashTag(tag));
-                }
+
+        if (isRenamingToLexicalTag) {
+            if (!getName().equals(tagNew)) {
+                setName(tagNew);
+                save();
+            }
+        } else {
+            // Create new tag if canonical tag doesn't already exist.
+            if (!TagUtils.hasCanonicalOfLexical(tagsBucket, tagNew)) {
+                TagUtils.createTag(tagsBucket, tagNew, index);
             }
 
-            // Add lexical tag to note.  Update this tag's name and save it.
-            if (isOldIdEqualToNewHash) {
-                tagsNew.add(tagNew);
-
-                if (!getName().equals(tagNew)) {
-                    setName(tagNew);
-                    save();
-                }
-            // Add new canonical tag to note and create new tag.  Delete this tag.
-            } else {
-                // Add new tag if note doesn't already have same hashed tag.
-                if (!tagsHash.contains(TagUtils.hashTag(tagNew))) {
-                    tagsNew.add(TagUtils.getCanonicalFromLexical(tagsBucket, tagNew));
-                }
-
-                // Create new tag if canonical tag doesn't already exist.
-                if (!TagUtils.hasCanonicalOfLexical(tagsBucket, tagNew)) {
-                    TagUtils.createTag(tagsBucket, tagNew, index);
-                }
-
-                delete();
+            delete();
+        }
+
+        ObjectCursor<Note> notesWithOldTag = findNotes(notesBucket, tagOld);
+
+        while (notesWithOldTag.moveToNext()) {
+            Note note = notesWithOldTag.getObject();
+            renameTagInNote(note, tagOld, tagNew, isRenamingToLexicalTag, tagsBucket);
+        }
+
+        notesWithOldTag.close();
+    }
+
+    private void renameTagInNote(Note note, String tagOld, String tagNew, boolean isRenamingToLexicalTag, Bucket<Tag> tagsBucket) {
+        List<String> tagsNew = new ArrayList<>();
+        List<String> tagsHash = new ArrayList<>();
+
+        // Create lists of note's tags excluding old tag.
+        for (String tag : note.getTags()) {
+            if (!tag.equals(tagOld)) {
+                tagsNew.add(tag);
+                tagsHash.add(TagUtils.hashTag(tag));
             }
+        }
 
-            // Add new tags to note and save it.
-            note.setTags(tagsNew);
-            note.save();
+        if (isRenamingToLexicalTag) {
+            tagsNew.add(tagNew);
+
+        } else {
+            // Add new tag if note doesn't already have a tag with the same hash.
+            if (!tagsHash.contains(TagUtils.hashTag(tagNew))) {
+                tagsNew.add(TagUtils.getCanonicalFromLexical(tagsBucket, tagNew));
+            }
```

Comment:
I don't think we have to introduce as much complexity in the note side; the logic should be straightforward for all updates regardless of whether we're changing the `tagHash` in the rename.